src/Storages/Hive/StorageHive.cpp

```diff
@@ -163,8 +166,26 @@ class StorageHiveSource : public SourceWithProgress, WithContext
                 std::unique_ptr<ReadBuffer> raw_read_buf;
                 try
                 {
-                    raw_read_buf = std::make_unique<ReadBufferFromHDFS>(
-                        hdfs_namenode_url, current_path, getContext()->getGlobalContext()->getConfigRef());
+                    auto get_raw_read_buf = [&]() -> std::unique_ptr<ReadBuffer>
+                    {
+                        auto buf = std::make_unique<ReadBufferFromHDFS>(
+                            hdfs_namenode_url, current_path, getContext()->getGlobalContext()->getConfigRef());
+
+                        bool thread_pool_read = read_settings.remote_fs_method == RemoteFSReadMethod::threadpool;
+                        if (thread_pool_read)
+                        {
+                            return std::make_unique<AsynchronousReadBufferFromHDFS>(StorageHive::getThreadPoolReader(), read_settings, std::move(buf));
+                        }
+                        else
+                        {
+                            return buf;
+                        }
+                    };
+                    raw_read_buf = get_raw_read_buf();
+                    if (read_settings.remote_fs_prefetch)
+                    {
+                        raw_read_buf->prefetch();
+                    }
```

Comment:
why does it give a noticeable improvement though? We read the result a few lines lower in reader->pull(), so not really obvious 
also for hive if there is cache, then we prefetch data here for no reason?

Reply:
Yes, its improvement is not obvious. I'll remove it

Reply:
done

Reply:
ok, but then why there is an improvement if just adding the pool if the we start waiting for the result right after putting the task into the threadpool?

Reply:
Because parallelism increased if we use read_method  `thread_pool`. 
In my environment, `max_threads` is 16, that's means parallel factor is 16 instead of 50 if `thread_pool` is not used.   

Besides, when read_method is `thread_pool`, ch would have the ability to limit global parallelism fetching from HDFS if multiple queries from different hive tables are running.